peloton/update_project.py

```diff
@@ -0,0 +1,1371 @@
+        login: str
+
+    @staticmethod
+    def _make_selector(
+        operation: sgqlc.operation.Operation
+    ) -> sgqlc.operation.Selector:
+        organization = operation.organization(login="SciTools")
+        team = organization.team(slug="Peloton")
+        return team.members
+
+    @staticmethod
+    def _extract_page(result: github_schema.Query) -> sgqlc.types.relay.Connection:
+        return result.organization.team.members
+
+    @staticmethod
+    def _extract_page_items(page: sgqlc.types.relay.Connection) -> list:
+        return page.nodes
+
+    def _add_query_fields(self, page) -> None:
+        nodes = page.nodes()
+        login = nodes.login()
+
+        if self.cols is None:
+            # This is run in a loop, but we only need to get the column names
+            #  once.
+            self.cols = self.ColNames.from_sgqlc_trees(
+                login=[login],
+            )
+
+    def _post_process(self) -> None:
+        super()._post_process()
+        # @rcomer has asked not to receive notifications for the
+        #  SciTools/peloton GitHub team, so is not a member and instead is
+        #  added here.
+        self.data_frame.loc[len(self.data_frame)] = (
+            {self.cols.login: "rcomer"}
+        )
+
+
+###############################################################################
+# MUTATIONS.
+
+
+class PaginatedMutation(abc.ABC):
+    """
+    Base class for sgqlc mutations, including pagination to avoid timeouts.
+    """
+
+    # _PAGINATION defines the number of mutation fields to include in one
+    #  mutation. If more are needed, the fields will be broken into
+    #  multiple pages.
+    _PAGINATION = 20
+    # All mutation fields include an `inputs` argument. The contents vary but
+    #  will always refer to the Peloton project.
+    _inputs_default = dict(project_id=PELOTON_PROJECT_ID)
+
+    def __init__(
+        self,
+        data_frame: pd.DataFrame,
+        project_cols: ProjectItemsQuery.ColNames,
+        issues_cols: IssuesQuery.ColNames,
+    ):
+        """
+        Execute the mutation as defined using the class methods.
+
+        Caller interaction with this instance is limited to accessing
+        self.result.
+        """
+        self._inputs = self._make_inputs(data_frame, project_cols, issues_cols)
+        self._indexes = list(range(len(self._inputs)))
+        # A mutation can only contain multiple 'fields' if they have different
+        #  names from each other.
+        self._aliases = [f"op_{ix}" for ix in self._indexes]
+
+        page_slices = [
+            slice(i, i + self._PAGINATION)
+            for i in range(0, len(self._inputs), self._PAGINATION)
+        ]
+        result_lists = []
+        for page_slice in page_slices:
+            result_lists.append(self._run_sub_mutation(page_slice))
+            # Avoid rate limiting from GitHub.
+            sleep(5)
+        self.result = self._get_data_from_sub_mutations(result_lists)
```

Comment:
@trexfeathers Could `result` be a read-only property to enforce the contract of the mutation API?

Reply:
How lovely 😊

aff11f2